File: Graph_traversal_algorithms/Topological_sort/topological_sort.py
from collections import deque             # For deque()

def topological_sort(adj_list):

    indegree = [0 for i in range(len(adj_list))]   
    
    vertices_with_indegree_zero = deque()
    
    visiting_order = []
    
    for each_vertex in adj_list:
    
        for each_neighbour in each_vertex:
        
            indegree[each_neighbour] += 1
            
    count = 0
    
    for each_vertex in indegree:
    
        if each_vertex == 0:
        
           vertices_with_indegree_zero.append(count)
           
        count += 1
           
    while (vertices_with_indegree_zero):
    
        vertex = vertices_with_indegree_zero.popleft()
        
        visiting_order.append(vertex)
        
        for each_neighbour in adj_list[vertex]:
        
            indegree[each_neighbour] -= 1
            
            if indegree[each_neighbour] == 0:
            
                vertices_with_indegree_zero.append(each_neighbour)
    
    if (visiting_order):
    
        print("\nThe visitng order is as follows:", end="\n")
    
    for vertex in visiting_order:
    
        print(vertex, end=" ")
    
    print("\n")        
                                

def check_vertex(vertex, nov):
    
    while(1):
        
        if (vertex >= nov or vertex < 0):
            
            print("Invalid vertex: ", vertex)
            
            vertex = int(input("Re-enter the vertex: "))
        
        else:
            return vertex


def read_edges(nov, noe, adj_list): 

    print("Enter the edges: ")

    while (noe):

        vertex = list(input().split())
        
        source, destination = int(vertex[0]), int(vertex[1])     
    
        source = check_vertex(source, nov)
        destination = check_vertex(destination, nov)
     
        adj_list[source].append(destination)
    
        noe -= 1 
           
    return adj_list

########################################################

no_of_vertices = int(input("Enter the no of vertices: "))

adj_list = [[] for i in range(0, no_of_vertices)]

no_of_edges = int(input("Enter the no of edges: "))

adj_list = read_edges(no_of_vertices, no_of_edges, adj_list)

topological_sort(adj_list)    

File: Graph_traversal_algorithms/Topological_sort/topological_sort.cpp
#include <iostream>
#include <vector>
#include <queue>


void topological_sort(const std::vector<std::vector<int>>& adj_list)
{

	std::vector<int> indegree(adj_list.size(), 0);

	std::queue<int> vertices_with_indegree_zero;

	std::vector<int> visiting_order;

	for (auto& neighbours: adj_list)
	{

		for (auto& neighbour_vertex:  neighbours)
			indegree[neighbour_vertex] += 1; 


	}

	int count = 0;

	for (auto& vertex : indegree)
	{

		if (vertex == 0)
			vertices_with_indegree_zero.push(count);

		count++;
	}

	while (!vertices_with_indegree_zero.empty())
	{

		int vertex_with_indegree_zero = vertices_with_indegree_zero.front();

		visiting_order.push_back(vertex_with_indegree_zero);

		vertices_with_indegree_zero.pop();

		for (auto v : adj_list[vertex_with_indegree_zero])
		{

			indegree[v] -= 1;

			if (indegree[v] == 0)
				vertices_with_indegree_zero.push(v);  

		}



	}

	if (visiting_order.size() != 0)
		std::cout << "\nThe visiting order of vertices is as follows:" << std::endl;

	for (auto& vertex : visiting_order)
		std::cout << vertex << " ";

	std::cout << std::endl;
}


void check_vertex(int& v, int& vc)	// check_vertex starts
{
	while (v >= vc || v < 0)
	{
		std::cout << "Invalid vertex no: " << v << std::endl;
		std::cout << "Re enter the vertex: " << std::endl;
		std::cin >> v;
	}
}	                               // check_vertex ends


void read_vertex_and_edge_count(int& v_count, int& e_count)
{

	std::cout << "Enter the no of vertices:" << std::endl;
	std::cin >> v_count;
	std::cout << "Enter the no of edges: " << std::endl;
	std::cin >> e_count;

}


void read_the_edges(std::vector<std::vector<int>>& adj, int& ec, int& v_count)
{

	int source;
	int destination;
	std::cout << "\nEnter the edges: " << std::endl;

	while (ec--)
	{
		std::cin >> source >> destination;

		check_vertex(source, v_count);
		check_vertex(destination, v_count);

		adj[source].push_back(destination);

	}

}


int main()
{

	int start_vertex;
	int vertex;
	int vertices_count = 0;
	int edges_count = 0;

	read_vertex_and_edge_count(vertices_count, edges_count);

	std::vector<std::vector<int>> adjacency_list(vertices_count);

	read_the_edges(adjacency_list, edges_count, vertices_count);

	topological_sort(adjacency_list);             // CALLING THE Topological Sort function

	return 0;

}	

